tests/testthat/test-subscript-tables.R

```diff
@@ -226,6 +226,190 @@ test_that("drop and exact recognised and used appropriately", {
+        tbl.vec <- as.vector(tbl)
+        unique.vals <- tbl.vec[!duplicated(tbl.vec) & !duplicated(rev(tbl.vec))]
+        chosen.val <- unique.vals[length(unique.vals)]
+        arr.idx <- drop(which(tbl == chosen.val, arr.ind = TRUE, useNames = FALSE))
+        if (!is.null(dim(arr.idx)) || NROW(arr.idx) == 0)  # every value in tbl appears more than once
+        { #  pick randomly from center
+            dims <- dim(tbl)
+            arr.idx <- sapply(dims, function(dim.len)
+                {
+                    if (dim.len == 1)
+                        return(1)
+                    sample(2:dim.len, 1)
+                })
+        }
+        out <- do.call(`[`, c(list(tbl), as.list(arr.idx)))
+        attr.zstat <- attr(out, "QStatisticsTestingInfo")[, "zstatistic"]
+        if (is.nan(attr.zstat))
+            attr.zstat <- NA_real_
+        expect_equal(as.numeric(out), attr.zstat, check.attributes = FALSE)
+        label.idx <- mapply(`[`, dimnames(tbl), arr.idx)
+        out <- do.call(`[`, c(list(tbl), as.list(label.idx)))
+        expect_equal(as.numeric(out), attr.zstat, check.attributes = FALSE)
+    }
+})
+
+## TODO: 1D Tables (char., int, logical idx),
+## 4D tables (multi x multi, grid x multi, multi x grid, grid x grid;
+## 3D output, 2D output, 1D output),
+## 5D Tables, 2D multistat tables,
+## int vector subset, D-column matrix subset, logical array subset,
+
+#####################################################################
+## 1D tables
+set.seed(23)
+dim.lens.avail <- vapply(lapply(tbls, dim), length, 1L)
+test.cases <- names(tbls)[dim.lens.avail == 1L]
+for (test.case in test.cases)
+{
+    tbl <- tbls[[test.case]]
+    start.idx <- sample(length(tbl), 1)
+    end.idx <- sample(seq_along(tbl)[-start.idx], 1)
+    test.name <- paste0("DS-3809: subvectors of 1D qTables: ",
+                        test.case, "[", start.idx, ":", end.idx,"]")
+    test_that(test.name,
+              {
+                  idx <- start.idx:end.idx
+                  z.stat.out <- attr(tbl[idx],
+                                     "QStatisticsTestingInfo")[, "zstatistic"]
+                  expected <- unclass(tbl)[idx]
+                  expected[is.nan(expected)] <- NA
+                  expect_equal(z.stat.out, expected, check.attributes = FALSE)
+
+                  char.idx <- names(tbl)[idx]
+                  z.stat.out <- attr(tbl[char.idx],
+                                     "QStatisticsTestingInfo")[, "zstatistic"]
+                  expected <- unclass(tbl)[char.idx]
+                  expected[is.nan(expected)] <- NA
+                  expect_equal(z.stat.out, expected, check.attributes = FALSE)
+
+                  logical.idx <- seq_along(tbl) %in% idx
+                  z.stat.out <- attr(tbl[logical.idx],
+                                     "QStatisticsTestingInfo")[, "zstatistic"]
+                  expected <- unclass(tbl)[logical.idx]
+                  expected[is.nan(expected)] <- NA
```

Comment:
More generally though, the `NaN` check should occur on the attribute, not the table. From what I gather the `NA` values are in the table when the z-statistic is `NaN` in the attribute. ie.g. There shouldn't be checks of ultimately is `is.nan(tbl[idx])` but instead should be `is.nan(attr.table)`

In the (current) suite of 1d test cases, there are no `NaN` values, so those reassignment lines can be removed.
```
  significancearrowratio significancedirection significancefontsizemultiplier significanceissignificant zstatistic
1                      1                    Up                       4.890000                      TRUE   5.044299
2                      1                    Up                       4.890000                      TRUE   5.161608
3                      1                  Down                       0.204499                      TRUE -10.205908
4                      1                    Up                       4.890000                      TRUE  25.573424
    pcorrected
1 4.551866e-07
2 2.448370e-07
3 1.865686e-24
4 0.000000e+00

$Ranking
  significancearrowratio significancedirection significancefontsizemultiplier significanceissignificant zstatistic
1              1.0000000                    Up                      4.8900000                      TRUE        Inf
2              0.2467866                  Down                      0.5102041                      TRUE  -2.401942
3              0.0000000                  None                      1.0000000                     FALSE   0.442783
4              0.0000000                  None                      1.0000000                     FALSE  -1.648969
5              1.0000000                  Down                      0.2044990                      TRUE       -Inf
6              0.0000000                  None                      1.0000000                     FALSE  -1.134204
    pcorrected
1 0.000000e+00
2 1.630828e-02
3 6.579227e-01
4 9.915391e-02
5 1.190338e-81
6 2.567089e-01
```

The check should be updated for the higher dim (2 or more) though.

Reply:
Not sure what you mean. It's checking for NaNs in `expected` and converting them to `NA` to obtain the desired expected output.

Reply:
The checks for `NaN` are unnecessary because `all.equal(NA_real_, NaN) == TRUE`. I removed them in c8a66fd8c44.